syz-verifier/verf/verifier.go

```diff
@@ -21,25 +21,67 @@ type Result struct {
 	Info ipc.ProgInfo
 }
 
+type ResultReport struct {
+	// Prog is the serialized program.
+	Prog string
+	// Mismatch contains the indexes of calls where mismatches were found.
+	Mismatch []int
+	// Reports contains information about each system call.
+	Reports []CallReport
+}
+
+type CallReport struct {
+	// Call is the name of the system call.
+	Call string
+	// Errno is a map between pools and the errno returned by executing the
+	// system call on a VM spawned by the respective pool.
+	Errnos map[int]int
+}
+
 // Verify checks whether the Results of the same program, executed on different
-// kernels are the same. If that's the case, it returns true, otherwise it
-// returns false.
-func Verify(res []*Result, prog *prog.Prog) bool {
+// kernels are the same. If that's not the case, it returns a ResultReport
+// which highlights the differences.
+func Verify(res []*Result, prog *prog.Prog) *ResultReport {
+	rr := &ResultReport{
+		Prog: string(prog.Serialize()),
+	}
+
+	c1 := res[0].Info.Calls
+	for idx, c := range c1 {
+		cr := CallReport{
+			Call:   prog.Calls[idx].Meta.Name,
+			Errnos: map[int]int{res[0].Pool: c.Errno},
+		}
+		rr.Reports = append(rr.Reports, cr)
+	}
+
+	send := false
 	for i := 1; i < len(res); i++ {
-		if !VerifyErrnos(res[0].Info.Calls, res[i].Info.Calls) {
-			return false
+		resi := res[i]
+		ci := resi.Info.Calls
+		if match, idx := VerifyErrnos(c1, ci); !match {
+			send = true
+			rr.Mismatch = append(rr.Mismatch, idx)
+		}
+		for j, c := range ci {
+			rr.Reports[j].Errnos[resi.Pool] = c.Errno
 		}
 	}
-	return true
+
+	if send {
```

Comment:
I could loop through the `ResultReport` and check whether, for any `CallReport`, `Mismatch` is set to `true`. Do you think that would be better?

Reply:
Or if we add a `Send` field in the `ResultReport`.

Reply:
Yes, I posted a conflicting comment that suggests to remove Mismatch (as a separate slice).

Reply:
> Or if we add a Send field in the ResultReport.

This feels wrong because `send` is only needed as a local variable in this function. For the callers the signal is non/non-nil result. Send will always be true for callers, which is confusing and not useful for anybody outside of this function.

Reply:
Ah, Marco was referring to the slice rather than the new field you suggested. Sorry, I misunderstood the comment.

> This feels wrong because send is only needed as a local variable in this function. For the callers the signal is non/non-nil result. Send will always be true for callers, which is confusing and not useful for anybody outside of this function.
I agree and I would prefer keeping it as a local variable

Reply:
Also, I'm sorry I haven't updated the PR yet, that might have made it more difficult for you to review as well. But I introduced a bug that took way more time than expected to spot.